stm32: Enable DMA MUX client for nucleo_wb55rg

FRASTM opened the pull request:
Initial work to support the spi as a client of the dmamux

test on nucleo_wb55rg board 
The DT is filled with parameters to set the peripheral request on the dmamux.
The dmamux channels used for the client transfer (rx, tx) are assigned by the DT among the total nb of dmamux channels (ex.14)


Signed-off-by: Francois Ramu francois.ramu@st.com

EDIT (erwango): SPI impacts removed from this PR (moved to #23157 that has been merged separately)

FRASTM commented:
The test application 'spi loopback' requires wired connection from MISO to MOSI of the SPI 1 on the nucleo_wb55rg. The SPI (client) transfer is using the dma and mux with request ID set to 7 and 6 as defined by the dts/arm/st/wb/stm32wb55Xg.dtsi
The test application 'spi loopback' does not consider half transfer or partial transfer when using the dma.

The SPI driver when activating the dma, is also controlling that the rx and/or tx buffers are valid before configuring the DMA Memory addresses (as a source or as a destination).

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -175,7 +185,66 @@ static int spi_null_tx_buf(struct device *dev, struct spi_config *spi_conf)
+}
+
+#ifdef CONFIG_SPI_STM32_DMA
+static int spi_null_rx_buf(struct device *dev, struct spi_config *spi_conf)
```
This test seems useless. You send  buffer_rx, which name is against what the test is doing, and verify that it's content is still what you have put inside?

If the test is supposed to test null_rx, it will be like testing spi_write(). No need for any rx related spi_buf/spi_buf_set (and it will not be STM32_DMA specific at all also)

FRASTM replied:
spi_null_rx_buf removed

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -193,23 +262,34 @@ static int spi_null_tx_buf(struct device *dev, struct spi_config *spi_conf)
+			.len = BUF_SIZE,
+#else
 			.len = 8,
+#endif
```
why inverting the lengths here? And why is that stm32 dma specific as well?

FRASTM replied:
spi_null_rx_buf removed

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -137,15 +140,22 @@ static int spi_complete_loop(struct device *dev, struct spi_config *spi_conf)
 static int spi_null_tx_buf(struct device *dev, struct spi_config *spi_conf)
 {
+#ifdef CONFIG_SPI_STM32_DMA
+/* if buffer address is null, the dma won't transfer */
```
You'll need to find a way to follow the API, which is transferring 0s if tx buffer is NULL. Have a static tx buffer in your driver meant for that use-case, something like that.

FRASTM replied:
this is now handled in the `drivers/spi/spi_ll_stm32.c`, directly in the `spi_stm32_dma_tx_load` function : 
-->   /* if tx buff is null, then sends NOP on the line. */

tbursztyka reviewed (commented):
You should not need to change the tests in spi_loopback: your driver needs to be ready to handle the various use-cases (or return -ENOTSUP if it is impossible to support some)

FRASTM commented:
this PR adds the spi as a client of the dmamux for the stm32wb55 series
on top of the https://github.com/zephyrproject-rtos/zephyr/pull/22021

FRASTM committed:
tests: drivers: spi loopback on nucleo_wb55 board

This configures the nucleo_wb55rg board for testing
the SPI loopback with DMA MUX transfer.
This test requires pin SPI_A_MISO (D12) to be connected
to pin SPI_A_MOSI (D11) on the nucleo_wb55rg board

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase

FRASTM commented:
changed the DTS for dma client of the SPI for nucleo_wb55
removed `compatible = "st,stm32-spi"; `  in the .overlay for this test

FRASTM committed:
soc: arm: stm32wb55x series enabling dma and dmamux

This patch defines the dma feature for the stm32wbXX
and the dmamux feature for the stm32wb55x
soc series from STMicroelectronics

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
remove the macro from the dt_fixup.h

FRASTM commented:
rebase after merge of https://github.com/zephyrproject-rtos/zephyr/pull/23157
Testcase with multiple loop on wb55rg through dmamux

erwango commented on soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series:
```diff
@@ -10,4 +10,15 @@ source "soc/arm/st_stm32/stm32wb/Kconfig.defconfig.stm32wb*"
 config SOC_SERIES
 	default "stm32wb"
 
+config I2C_STM32_V2
+	default y
+	depends on I2C_STM32
```
Remove these lines

FRASTM replied:
ok

erwango commented on soc/arm/st_stm32/stm32wb/Kconfig.defconfig.stm32wb55xx:
```diff
@@ -11,4 +11,11 @@ config SOC
 
+if DMA_STM32
+
+config DMAMUX_STM32
```
Move to soc/arm/st_stm32/stm32wb/Kconfig.defconfig.series

FRASTM replied:
keeping dmamux feature for this serie only

erwango reviewed (changes requested):
Minor comment otherwise LGTM

FRASTM commented:
changing syntax in Kconfig for stm32wbxx

FRASTM commented:
rebase after merge of https://github.com/zephyrproject-rtos/zephyr/pull/22021

galak commented:
This needs rebasing to deal with merge conflict.

FRASTM committed:
dts: arm: stm32wb: configure DMA and DMAMUX

This patch enables the DMA and DMAMUX on the stm32wb55x.
The DMAMUX has 14 channels.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase and fix merge conflict

carlescufi commented:
@galak could you merge this one if happy?
